File: frontend-react/src/lib/api/monitoring.ts
/** Monitoring API methods */
import { ApiClient } from './client';
import type { MonitoringOverview, MonitoringSyncLogEntry } from '@/types/api';

export function addMonitoringMethods(api: ApiClient) {
    return {
        getMonitoringOverview() {
            return api.request<MonitoringOverview>('GET', '/api/v1/monitoring/overview');
        },
        getMonitoringSyncLog(params?: {
            service?: string;
            sync_type?: string;
            status?: string;
            limit?: number;
            offset?: number;
        }) {
            const searchParams = new URLSearchParams();
            if (params?.service) searchParams.set('service', params.service);
            if (params?.sync_type) searchParams.set('sync_type', params.sync_type);
            if (params?.status) searchParams.set('status', params.status);
            if (params?.limit) searchParams.set('limit', String(params.limit));
            if (params?.offset) searchParams.set('offset', String(params.offset));
            const qs = searchParams.toString();
            return api.request<MonitoringSyncLogEntry[]>(
                'GET',
                `/api/v1/monitoring/sync_log${qs ? '?' + qs : ''}`,
            );
        },
    };
}

File: backend/routers/monitoring.py
"""
Router: /monitoring — sync health dashboard, scheduler status.
Thin HTTP layer — logic in services/monitoring_service.py.
"""

import logging

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.models import Project
from backend.project_context import get_current_project
from backend.schemas.monitoring import (
    MonitoringOverview,
    SchedulerJobInfo,
    SchedulerStatus,
    SyncLogEntry,
    SyncTypeStatus,
)
from backend.services import monitoring_service

logger = logging.getLogger("dds.monitoring")

router = APIRouter(prefix="/monitoring")


@router.get("/overview", response_model=MonitoringOverview)
async def get_overview(
    project: Project = Depends(get_current_project),
    db: AsyncSession = Depends(get_db),
):
    """Get monitoring overview: sync health + scheduler status."""
    overview = await monitoring_service.get_sync_overview(db, project.id)

    # Get scheduler info — try in-memory first (works in worker),
    # fall back to DB-based detection (works in api container)
    from backend.scheduler import get_scheduler_info

    sched_info = get_scheduler_info()

    # If in-memory scheduler is not available (api container),
    # infer status from recent sync_log activity
    if not sched_info["running"] and overview["total_syncs_24h"] > 0:
        sched_info = {"running": True, "jobs": []}
        logger.debug("Scheduler detected as running via sync_log activity (api container)")

    return MonitoringOverview(
        sync_types=[SyncTypeStatus(**st) for st in overview["sync_types"]],
        scheduler=SchedulerStatus(
            running=sched_info["running"],
            jobs=[SchedulerJobInfo(**j, is_active=True) for j in sched_info["jobs"]],
        ),
        total_syncs_24h=overview["total_syncs_24h"],
        total_errors_24h=overview["total_errors_24h"],
    )


@router.get("/sync_log", response_model=list[SyncLogEntry])
async def get_sync_log(
    service: str | None = Query(None),
    sync_type: str | None = Query(None),
    status: str | None = Query(None),
    limit: int = Query(50, le=200),
    offset: int = Query(0),
    project: Project = Depends(get_current_project),
    db: AsyncSession = Depends(get_db),
):
    """Get filtered sync log with duration calculation."""
    return await monitoring_service.get_sync_log_filtered(
        db,
        project.id,
        service,
        sync_type,
        status,
        limit,
        offset,
    )
